### PR TITLE
build: add validation to the specified dockerfile path

### LINKED ISSUE #1104: [RFE?] `buildah bud -f <directory>` causes panic
**Description**

When using `buildah bud` with an explicit Dockerfile (`-f`), it'd be wonderful to handle directories nicely. This is helpful in, e.g., scripting where sometimes the user's path is incorrect or they fat-fingered a variable. :)

**Steps to reproduce the issue:**

Consider the difference in the following commands:

```bash
$ buildah bud --tag tag -f "tools/Dockerfiles/stylecheck" .
<does something useful>
```

```bash
$ buildah bud --tag tag -f "tools/Dockerfiles/NOTAFILE" .
error reading "/home/cipherboy/GitHub/sandbox/jss/tools/Dockerfiles/NOTAFILE": open /home/cipherboy/GitHub/sandbox/jss/tools/Dockerfiles/NOTAFILE: no such file or directory
ERRO[0000] exit status 1
```

But in the event that the file isn't specified or happens to be a directory, we get the cryptic panic:
```bash
$ buildah bud -f "tools/Dockerfiles/" .
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/containers/buildah/imagebuildah.(*Executor).Build(0xc4200e4dc0, 0xd32580, 0xc420028128, 0x0, 0x0, 0x0, 0x0, 0x0)
	/builddir/build/BUILD/buildah-608fa843cce45e7ee58ccb71a90297b645a984d3/_build/src/github.com/containers/buildah/imagebuildah/build.go:1185 +0x670
github.com/containers/buildah/imagebuildah.BuildDockerfiles(0xd32580, 0xc420028128, 0xd41f00, 0xc420414480, 0xc42030e150, 0x22, 0x0, 0x0, 0x0, 0x0, ...)
	/builddir/build/BUILD/buildah-608fa843cce45e7ee58ccb71a90297b645a984d3/_build/src/github.com/containers/buildah/imagebuildah/build.go:1284 +0xfe7
main.budCmd(0xc4200e06e0, 0x0, 0x0)
	/builddir/build/BUILD/buildah-608fa843cce45e7ee58ccb71a90297b645a984d3/cmd/buildah/bud.go:236 +0x17ac
github.com/urfave/cli.HandleAction(0xb76160, 0xcc2010, 0xc4200e06e0, 0x0, 0xc420388d20)
	/builddir/build/BUILD/buildah-608fa843cce45e7ee58ccb71a90297b645a984d3/src/github.com/urfave/cli/app.go:501 +0xc8
github.com/urfave/cli.Command.Run(0xc8ea7d, 0x16, 0x0, 0x0, 0x1288190, 0x1, 0x1, 0xcaaa89, 0x31, 0x0, ...)
	/builddir/build/BUILD/buildah-608fa843cce45e7ee58ccb71a90297b645a984d3/src/github.com/urfave/cli/command.go:165 +0x47d
github.com/urfave/cli.(*App).Run(0xc4202161c0, 0xc42001e0b0, 0xb, 0xb, 0x0, 0x0)
	/builddir/build/BUILD/buildah-608fa843cce45e7ee58ccb71a90297b645a984d3/src/github.com/urfave/cli/app.go:259 +0x6e8
main.main()
	/builddir/build/BUILD/buildah-608fa843cce45e7ee58ccb71a90297b645a984d3/cmd/buildah/main.go:120 +0xe22
ERRO[0000] exit status 2
```

**Describe the results you received:**

A panic! :o

**Describe the results you expected:**

Something vaguely useful like "Dockerfile is a directory", which would be better than Docker's results:

```bash
$ docker build -f "tools/Dockerfiles/" .
Sending build context to Docker daemon 20.87 MB
Error response from daemon: No image was generated. Is your Dockerfile empty?
```

(the above is regardless whether or not the directory contains a file named Dockerfile)

**Output of `rpm -q buildah` or `apt list buildah`:**

```
buildah-1.4-3.git608fa84.fc28.x86_64
```

**Output of `buildah version`:**

```
Version:         1.4
Go Version:      go1.10.4
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:  v0.7.0-alpha1
Git Commit:      608fa84
Built:           Thu Oct  4 12:40:10 2018
OS/Arch:         linux/amd64
```

(skipping the rest of the information...).

### PR DESCRIPTION
**Pull request by q384566678**
Make sure the path to the specified dockerfile is a file, excluding the folder.

fixes #1104 

After change:
```
➜  ~ buildah bud -f "tools/Dockerfiles" .
"tools/Dockerfiles" is a directory
ERRO[0000] exit status 1 
```

Signed-off-by: Zhou Hao <zhouhao@cn.fujitsu.com>

### CODE REVIEW

**Comment by q384566678** [comment]
@cipherboy PTAL

**Comment by cipherboy** [comment]
@q384566678 LGTM; though I didn't figure out how to build `buildah` from source... :) I think I'm missing a lot of dependencies that `dnf builddep buildah` didn't pull in. I'll leave the (formal) review to someone else.

But yes, thank you very much! The new error messages look nice. :)

File: imagebuildah/build.go [hunk 1]
```diff
@@ -1289,9 +1289,18 @@ func BuildDockerfiles(ctx context.Context, store storage.Store, options BuildOpt
...
 			// If the Dockerfile isn't found try prepending the
 			// context directory to it.
-			if _, err := os.Stat(dfile); os.IsNotExist(err) {
-				dfile = filepath.Join(options.ContextDirectory, dfile)
+			s, err := os.Stat(dfile)
+			if err != nil {
+				if os.IsNotExist(err) {
+					dfile = filepath.Join(options.ContextDirectory, dfile)
+				} else {
+					return "", nil, errors.Wrapf(err, "error reading %q", dfile)
```
Comment (rhatdan) [inline imagebuildah/build.go]: how about "Unable to stat file %q

File: imagebuildah/build.go [hunk 2]
```diff
@@ -1289,9 +1289,18 @@ func BuildDockerfiles(ctx context.Context, store storage.Store, options BuildOpt
...
+				}
 			}
+			if s.IsDir() {
```
Comment (rhatdan) [inline imagebuildah/build.go]: Should check to make sure it is a file.  Could be a pipe, socket, chr_file, ...

**Comment by TomSweeneyRedHat** [comment]
Test issues look to be unrelated.  A number of segv's trying to deal with vfs.

File: imagebuildah/build.go [hunk 3]
```diff
@@ -1289,9 +1289,18 @@ func BuildDockerfiles(ctx context.Context, store storage.Store, options BuildOpt
 		} else {
 			// If the Dockerfile isn't found try prepending the
 			// context directory to it.
-			if _, err := os.Stat(dfile); os.IsNotExist(err) {
-				dfile = filepath.Join(options.ContextDirectory, dfile)
+			s, err := os.Stat(dfile)
+			if err != nil {
+				if os.IsNotExist(err) {
+					dfile = filepath.Join(options.ContextDirectory, dfile)
```
[the remark this answers is not in the document]
Comment (q384566678) [reply imagebuildah/build.go]: We need to get the `FileInfo` through `Stat` and then determine if it is a file.

**Comment by rhatdan** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
@TomSweeneyRedHat @q384566678 Does #1132 remove the need for this PR?

**Comment by TomSweeneyRedHat** [comment]
Oops, yep, #1132 replaces this.  Sorry @q384566678 , I'd forgotten you'd this one in progress and the issue didn't have anyone assigned to it.  I'm going to close this, thanks for the work @q384566678 !

**Pull request closed without merging**